pinot-broker/src/main/java/com/linkedin/pinot/broker/queryquota/TableQueryQuotaManager.java

```diff
@@ -0,0 +1,138 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.broker.queryquota;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.RateLimiter;
+import com.linkedin.pinot.common.config.QuotaConfig;
+import com.linkedin.pinot.common.config.TableConfig;
+import com.linkedin.pinot.common.config.TableNameBuilder;
+import com.linkedin.pinot.common.metadata.ZKMetadataProvider;
+import com.linkedin.pinot.common.utils.CommonConstants;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.annotation.Nonnull;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.model.ExternalView;
+import org.apache.helix.store.zk.ZkHelixPropertyStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class TableQueryQuotaManager {
+  private static final Logger LOGGER = LoggerFactory.getLogger(TableQueryQuotaManager.class);
+
+  private final ZkHelixPropertyStore<ZNRecord> _propertyStore;
+  private final Map<String, RateLimiter> _rateLimiterMap;
+
+  public TableQueryQuotaManager(ZkHelixPropertyStore<ZNRecord> propertyStore) {
+    _propertyStore = propertyStore;
+    _rateLimiterMap  = new ConcurrentHashMap<>();
+  }
+
+  /**
+   * Initialize dynamic rate limiter with table query quota.
+   * @param tableConfig table config.
+   * @param brokerResource broker resource which stores all the broker states of each table.
+   * */
+  public void initTableQueryQuota(TableConfig tableConfig, ExternalView brokerResource) {
+    String tableName = tableConfig.getTableName();
+    String rawTableName = TableNameBuilder.extractRawTableName(tableName);
+    LOGGER.info("Initializing rate limiter for table {}", rawTableName);
+
+    CommonConstants.Helix.TableType tableType = TableNameBuilder.getTableTypeFromTableName(tableName);
+    if (tableType == CommonConstants.Helix.TableType.OFFLINE) {
+      QuotaConfig quotaConfig = tableConfig.getQuotaConfig();
+      if (quotaConfig != null && quotaConfig.getQps() != null) {
+        createRateLimiter(tableName, rawTableName, brokerResource, quotaConfig);
+      } else {
+        String realtimeTableName = TableNameBuilder.REALTIME.tableNameWithType(rawTableName);
+        TableConfig realTimeTableConfig = ZKMetadataProvider.getTableConfig(_propertyStore, realtimeTableName);
+        if (realTimeTableConfig == null) {
+          return;
+        }
+        QuotaConfig realTimeTableQuotaConfig = realTimeTableConfig.getQuotaConfig();
```

Comment:
That's because we want to reuse the `QuotaConfig` class by adding `qps` field in it, while `QuotaConfig` field is under `TableConfig`. And `TableConfig` belongs to the table with a specific type(real-time or offline).  We can discuss it to find a proper place to store this `qps` field.